File: No_2491/Divide Players Into Teams of Equal Skill.py
class Solution:
    def dividePlayers(self, skill: List[int]) -> int:
        target = sum(skill) // (len(skill) // 2)
        freq = defaultdict(int)
        ret = 0
        
        for num in skill:
            gap = target - num
            if gap in freq:
                freq[gap] -= 1
                if freq[gap] == 0:
                    freq.pop(gap)
                ret += num * gap
            else:
                freq[num] += 1
        
        return ret if not freq else -1

File: No_2491/Divide Players Into Teams of Equal Skill.cpp
class Solution {
public:
    long long dividePlayers(vector<int>& skill) {
        long long ret = 0;
        int target = accumulate(skill.begin(), skill.end(), 0) / (skill.size() / 2);
        unordered_map<int, int> hmap;
        
        for (int num : skill) {
            int gap = target - num;
            if (hmap.find(gap) != hmap.end()) {
                hmap[gap] -= 1;
                if (hmap[gap] == 0)
                    hmap.erase(gap);
                ret += num * gap;
            } else {
                hmap[num]++;
            }
        }

        return hmap.size() == 0 ? ret : -1;
    }
};